### ISSUE TITLE
Type confusion and OOB memory access in positional record structs 

### ISSUE THREAD

**Issue by C-xC-c** [dotnet/roslyn#59010]
In record structs with positional parameters, if you create a constructor which in turn calls the synthesised parameterless constructor, the compiler will emit code which either attempts an out of bounds memory access, or results in a type confusion. 

The OOB read is caused when the positional parameters outnumber the constructor's arguments. The below code will compile, and throw an `System.InvalidProgramException` at runtime.
```cs
record struct Parameters(int i, int j)
{
    public Parameters(int i)
        : this()
    {
    }
}
```
And the generated IL:
```
.method public hidebysig specialname rtspecialname 
	instance void .ctor (
		int32 i
	) cil managed 
{
	// Method begins at RVA 0x2094
	// Header size: 1
	// Code size: 15 (0xf)
	.maxstack 8

	IL_0000: ldarg.0
	IL_0001: ldarg.1
	IL_0002: stfld int32 Parameters::'<i>k__BackingField'
	IL_0007: ldarg.0
	IL_0008: ldarg.2 /*Error near IL_0008: ldarg 2 (out-of-bounds)*/
	IL_0009: stfld int32 Parameters::'<j>k__BackingField'
	IL_000e: ret
} // end of method Parameters::.ctor
```
More concerning, if the number of arguments are equal the constructor will write the values to the corresponding positional parameters, regardless of type; essentially a bitwise conversion. This will compile and run to completion, despite invoking undefined (but reproducible) behaviour.
```cs
record struct TypeConfusion(ulong i)
{
    public TypeConfusion(string s)
        : this()
    {
    }
}

string s = "test";

 // Confusion between the string's pointer to ulong; the address of the string is now held in tc.i
TypeConfusion tc = new(s);

// We can prove this by comparing against the address of s[0].
fixed (char* c = s.AsSpan())
{
    Console.WriteLine(((ulong)c) - 12 == tc.i); // -12 to account for the object header.
}
```

And the constructor's IL:
```
.method public hidebysig specialname rtspecialname 
	instance void .ctor (
		string s
	) cil managed 
{
	// Method begins at RVA 0x20b4
	// Header size: 1
	// Code size: 8 (0x8)
	.maxstack 8

        // i = (ulong)(long)s;
	IL_0000: ldarg.0
	IL_0001: ldarg.1
	IL_0002: stfld uint64 TypeConfusion::'<i>k__BackingField'
	IL_0007: ret
} // end of method TypeConfusion::.ctor
```
This could potentially open the door to many security vulnerabilities, including possible ACE.

**Version Used**: 
```
$ dotnet --version
6.0.101
```

**Comment by cston** [comment]
@C-xC-c, thanks for reporting this issue.

This is https://github.com/dotnet/roslyn/issues/58328, fixed in https://github.com/dotnet/roslyn/pull/58339.

**Comment by notflan** [comment]
Hi @cston, could you help clarify some things for me regarding this?

> This is #58328, fixed in #58339.

In the linked issue, the reporter says the field is filled with garbage or uninitialised memory, but it's actually a pointer to the managed object's header itself.
Is it not a little concerning that you can use this bug to cause the runtime to arbitrarily interpret any reference type as any other reference type, or reinterpret any arbitrary memory address passed as a pointer-sized integer to a managed object of any type, using no `unsafe` code at all?

Could this fix be backported to the latest stable release? Because from what I can see, it seems to me more serious than just getting crashes and garbage values.
I'm not an expert, but from my observations with this bug it can be used, either accidentally or intentionally, to do very unsafe operations with no `unsafe` code blocks or functions rather trivially.

A benign example would be something to the effect of:
```cs
public static class ObjectPointerExt
{
    private record struct ObjectPointer(ulong Pointer)
    {
        public ObjectPointer(object o) : this() { }
    }
    private record struct PointerObject(object Obj)
    {
        public PointerObject(ulong ptr) : this() { }
    }

    public static UIntPtr GetPointerTo(this object o) => (UIntPtr) (new ObjectPointer(o).Pointer);
    public static object GetObjectFrom(this UIntPtr p) => new PointerObject((ulong)p).Obj;
}
```
Which wouldn't require the project to allow `unsafe` code. (I haven't tested this with generics but I don't see why it wouldn't work with any type `where T: class` as well.)
Bitwise conversions between any value types should also be possible although I haven't tested that with a user-defined value type.

There may be potential to use this to modify or swap object vtables, headers, or any part of a managed object without needing any `unsafe` code. As well as reinterpret and modify any selected memory address as any type, user-defined or not.

I'm not sure if this would open up any real-world exploit opportunities in codebases where this bug is undetected, but if patching the fix in to the latest stable release isn't possible would adding a compiler warning for instances of this pattern in codebases which have it be possible? I had a codebase which had this bug in and would've gone undetected if not for the fact the number of constructor arguments and positional parameters were different for the specific constructor that started causing mysterious crashes. The rest of the constructors were silently reinterpreting the fields without my noticing.

Thanks.
